README.md
# Supported Browsers

> Matrix of supported browsers for automated javascript testing

[![npm version](https://badge.fury.io/js/%40ideal-postcodes%2Fsupported-browsers.svg)](https://badge.fury.io/js/%40ideal-postcodes%2Fsupported-browsers)
![Release](https://github.com/ideal-postcodes/supported-browsers/workflows/Release/badge.svg)
![CI](https://github.com/ideal-postcodes/supported-browsers/workflows/CI/badge.svg)
![Cross Browser Testing](https://github.com/ideal-postcodes/supported-browsers/workflows/Cross%20Browser%20Testing/badge.svg)

`@ideal-postcodes/supported-browsers` exports a matrix of supported browsers

## Documentation

Browser exports for [Saucelabs](https://saucelabs.com)

### Supported Browsers

#### Latest Desktop

- Chrome
- Safari
- Firefox
- Edge

#### Latest Mobile

- Safari (iOS)
- Chrome (Android)

#### Legacy Desktop

- Internet Explorer 11

#### Legacy Mobile

- Chrome (Android 6.0)
- Safari (iOS 10.3)

### Configuration & Usage

```bash
npm install @ideal-postcodes/supported-browsers
```

```javascript
// Retrieve tested browser matrix
import { browsers } from "@ideal-postcodes/supported-browsers";
```

## Test

```bash
npm test
```

## Licence

MIT


CHANGELOG.md
# [2.5.0](https://github.com/ideal-postcodes/supported-browsers/compare/2.4.0...2.5.0) (2020-11-27)


### Features

* **Android:** Bump legacy to 6.0 ([bbbe48a](https://github.com/ideal-postcodes/supported-browsers/commit/bbbe48a))

# [2.4.0](https://github.com/ideal-postcodes/supported-browsers/compare/2.3.2...2.4.0) (2020-11-24)


### Features

* **Saucelabs:** Update browser lists ([2176160](https://github.com/ideal-postcodes/supported-browsers/commit/2176160))

## [2.3.2](https://github.com/ideal-postcodes/supported-browsers/compare/2.3.1...2.3.2) (2020-05-26)


### Bug Fixes

* **Safari:** Set OS to Catalina ([b5e692b](https://github.com/ideal-postcodes/supported-browsers/commit/b5e692b))

## [2.3.1](https://github.com/ideal-postcodes/supported-browsers/compare/2.3.0...2.3.1) (2020-05-26)


### Bug Fixes

* **Config:** Enable singleRun ([dda38cb](https://github.com/ideal-postcodes/supported-browsers/commit/dda38cb))

# [2.3.0](https://github.com/ideal-postcodes/supported-browsers/compare/2.2.0...2.3.0) (2020-05-26)


### Features

* **Saucelabs:** Enable EU region ([78d9ea4](https://github.com/ideal-postcodes/supported-browsers/commit/78d9ea4))

# [2.2.0](https://github.com/ideal-postcodes/supported-browsers/compare/2.1.1...2.2.0) (2020-02-22)


### Features

* **Latest Mobile:** Update versions ([72706e7](https://github.com/ideal-postcodes/supported-browsers/commit/72706e7))

## [2.1.1](https://github.com/ideal-postcodes/supported-browsers/compare/2.1.0...2.1.1) (2020-02-20)


### Bug Fixes

* **Plugins:** Drop karma polyfill ([769a201](https://github.com/ideal-postcodes/supported-browsers/commit/769a201))

# [2.1.0](https://github.com/ideal-postcodes/supported-browsers/compare/2.0.0...2.1.0) (2020-02-20)


### Features

* **Util:** Add util functions to generate sauce config ([4678f53](https://github.com/ideal-postcodes/supported-browsers/commit/4678f53))

# [2.0.0](https://github.com/ideal-postcodes/supported-browsers/compare/1.2.1...2.0.0) (2020-02-19)


### Features

* **Saucelabs:** Migrate to saucelabs ([f13416f](https://github.com/ideal-postcodes/supported-browsers/commit/f13416f))


### BREAKING CHANGES

- CBT configuration removed
- Saucelabs configuration added
- Some mobile legacy floor versions bumped

## [1.2.1](https://github.com/ideal-postcodes/supported-browsers/compare/1.2.0...1.2.1) (2019-10-28)


### Bug Fixes

* **Safari:** Remove Safari Desktop ([54ca227](https://github.com/ideal-postcodes/supported-browsers/commit/54ca227))

# [1.2.0](https://github.com/ideal-postcodes/supported-browsers/compare/1.1.1...1.2.0) (2019-10-16)


### Features

* **ios:** Switch to iphone 6 simulator ([b37d76f](https://github.com/ideal-postcodes/supported-browsers/commit/b37d76f))

## [1.1.1](https://github.com/ideal-postcodes/supported-browsers/compare/1.1.0...1.1.1) (2019-10-15)


### Bug Fixes

* **CI:** Run latest node ([39f3bdd](https://github.com/ideal-postcodes/supported-browsers/commit/39f3bdd))

# [1.1.0](https://github.com/ideal-postcodes/supported-browsers/compare/1.0.0...1.1.0) (2019-10-14)


### Features

* **CBT:** Export CBT launcher helper ([c5ad9e2](https://github.com/ideal-postcodes/supported-browsers/commit/c5ad9e2))

# 1.0.0 (2019-10-03)


### Features

* **Browsers:** Migrate to CBT ([5801629](https://github.com/ideal-postcodes/supported-browsers/commit/5801629))


test/test.ts
import { assert } from "chai";
import {
  browsers,
  latestMobile,
  legacyMobile,
  latestDesktop,
  legacyDesktop,
  gitSha,
  ci,
  generateBuild,
  config,
} from "../lib/index";

describe("Support matrix", () => {
  it("exports a list of browsers", () => {
    assert.isTrue(Object.values(browsers).length > 0);
  });
  it("exports list of modern desktop browsers", () => {
    assert.isTrue(Object.values(latestDesktop).length > 0);
  });
  it("exports list of legacy desktop browsers", () => {
    assert.isTrue(Object.values(legacyDesktop).length > 0);
  });
  it("exports list of modern mobile browsers", () => {
    assert.isTrue(Object.values(latestMobile).length > 0);
  });
  it("exports list of legacy desktop browsers", () => {
    assert.isTrue(Object.values(legacyMobile).length > 0);
  });
});

describe("sauceConfig", () => {
  it("generates sauce config", () => {
    const conf = config({ testName: "foo" });
    assert.isString(conf.sauceLabs.build);
    assert.equal(conf.sauceLabs.region, "eu");
    assert.isTrue(conf.singleRun);
    assert.equal(conf.sauceLabs.testName, "foo");
  });
});

describe("generateBuild", () => {
  let env: any;

  before(() => {
    env = process.env;
  });

  after(() => {
    process.env = env;
  });

  it("genreates git sha if no CI", () => {
    process.env = { GITHUB_ACTION: undefined };
    assert.include(generateBuild(), gitSha());
  });
  it("generates GITHUB RUN ID if CI", () => {
    process.env = { GITHUB_ACTION: "foo", GITHUB_RUN_ID: "42" };
    assert.equal(generateBuild(), process.env.GITHUB_RUN_ID);
  });
});

describe("gitSha", () => {
  it("returns gitsha of current repo", () => {
    assert.isString(gitSha());
  });
});

describe("ci", () => {
  let env: any;

  before(() => {
    env = process.env;
  });

  after(() => {
    process.env = env;
  });

  it("returns true if CI detected", () => {
    process.env = { GITHUB_ACTION: "foo" };
    assert.isTrue(ci());
  });
  it("returns false if no CIdetected", () => {
    process.env = { GITHUB_ACTION: undefined };
    assert.isFalse(ci());
  });
});


test/null-test.ts
describe("Null test", () => {
  it("verifies that tests can run in the remote browser environment", () => {
    // @ts-ignore
    if ("P" === "NP") throw Error("Oh my!");
  });
});


lib/index.ts
import { execSync } from "child_process";
import { networkInterfaces } from "os";

export type SupportMatrix = Record<string, Browser>;

export type Browser = MobileBrowser | DesktopBrowser;

interface SauceOptions {
  screenResolution?: string;
}

interface DesktopBrowser {
  base: "SauceLabs";
  // Browser (e.g. IE, Chrome)
  browserName: string;
  // Browser version number and? bit version (e.g. 74x64)
  browserVersion: string;
  // Operating system and version (e.g. Windows 10)
  platformName?: string;
  "sauce:options"?: SauceOptions;
}

interface MobileBrowser {
  base: "SauceLabs";
  appiumVersion?: string;
  // Browser (e.g. IE, Chrome)
  browserName: string;
  // Browser version number and? bit version (e.g. 74x64)
  browserVersion?: string;
  // Operating system
  platformName?: string;
  platform?: string;
  // Operating system version number
  platformVersion?: string;
  version?: string;
  // Mobile device (e.g. iPhone 4s)
  deviceName?: string;
  device?: string;
  // Portrait or landscape
  deviceOrientation?: string;
}

const base = "SauceLabs";

const appiumVersion = "1.17.1";

export const latestDesktop: SupportMatrix = {
  chrome: {
    base,
    browserName: "chrome",
    browserVersion: "latest",
  },
  firefox: {
    base,
    browserName: "firefox",
    browserVersion: "latest",
  },
  safari: {
    base,
    browserName: "safari",
    browserVersion: "latest",
    platformName: "macOS 10.15",
  },
  edge: {
    base,
    browserName: "microsoftedge",
    browserVersion: "latest",
  },
};

export const legacyDesktop: SupportMatrix = {
  ie: {
    base,
    browserName: "internet explorer",
    browserVersion: "11",
    platformName: "Windows 8.1",
  },
};

export const latestMobile: SupportMatrix = {
  "ios-latest": {
    base,
    appiumVersion,
    deviceName: "iPhone 11 Simulator",
    platformVersion: "13.4",
    platformName: "iOS",
    browserName: "Safari",
  },
  "android-latest": {
    base,
    appiumVersion,
    deviceName: "Android GoogleAPI Emulator",
    browserName: "Chrome",
    platformVersion: "11.0",
    platformName: "Android",
  },
};

export const legacyMobile: SupportMatrix = {
  "android-legacy": {
    base,
    appiumVersion,
    deviceName: "Android GoogleAPI Emulator",
    browserName: "Chrome",
    platformVersion: "6.0",
    platformName: "Android",
  },
  "ios-legacy": {
    base,
    browserName: "Safari",
    deviceName: "iPhone Simulator",
    deviceOrientation: "portrait",
    platformVersion: "10.3",
    platformName: "iOS",
    appiumVersion: "1.9.1",
  },
};

/**
 * Export supported browsers
 */
export const browsers: SupportMatrix = {
  ...latestDesktop,
  ...legacyDesktop,
  ...latestMobile,
  ...legacyMobile,
};

/**
 * Return true if CI environment (Github actions) detected
 */
export const ci = (): boolean => process.env.GITHUB_ACTION !== undefined;

/**
 * Retrieve git sha
 */
export const gitSha = (): string =>
  execSync("git rev-parse --short HEAD").toString().trim();

/**
 * Build ID
 *
 * Use gitsha and date if local
 *
 * Use Github action ID if CI
 */
export const generateBuild = (): string =>
  ci() ? `${process.env.GITHUB_RUN_ID}` : `${gitSha()}-${new Date().toJSON()}`;

interface Options {
  defaults?: any;
  testName: string;
  build?: string;
}

const ni = networkInterfaces();
const hostname = Object.keys(ni)
  .map((interf) =>
    // @ts-ignore
    ni[interf].map((o) => !o.internal && o.family === "IPv4" && o.address)
  )
  .reduce((a, b) => a.concat(b))
  // @ts-ignore
  .filter((o) => o)[0];

/**
 * Generates sauce config
 */
export const config = ({ testName, build, defaults = {} }: Options) => ({
  ...defaults,
  // Hostname points to IP address. Due to a change in Android where the VM
  // will point to its own localhost and not karma server
  hostname,
  reporters: ["dots", "karma-typescript", "saucelabs"],
  plugins: ["karma-mocha", "karma-typescript", "karma-sauce-launcher"],
  browserDisconnectTimeout: 10000,
  browserDisconnectTolerance: 2,
  browserNoActivityTimeout: 30000,
  singleRun: true,
  captureTimeout: 120000,
  transports: ["websocket", "polling"],
  sauceLabs: {
    testName,
    region: "eu",
    startConnect: true,
    build: build === undefined ? generateBuild() : build,
    recordVideo: false,
    recordScreenshots: false,
    public: "public restricted",
  },
});
